### PR TITLE
Accurately implement steady & system clocks

### PR DESCRIPTION
**Pull request by Thog**
All 1.0.0 time cmds should be implemented with this.

### CODE REVIEW

File: Ryujinx.HLE/HOS/Services/Time/Clock/ClockTypes.cs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        public ulong ToSeconds()
+        {
+            return NanoSeconds / 1000000000;
+        }
+
+        public static TimeSpanType FromTicks(ulong ticks, ulong frequency)
+        {
+            return new TimeSpanType(ticks * 1000000000 / frequency);
+        }
+
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Services/Time/Clock/ClockTypes.cs]: Remove extra empty line

File: Ryujinx.HLE/HOS/Services/Time/Clock/ClockTypes.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    [StructLayout(LayoutKind.Sequential)]
+    public struct SteadyClockTimePoint
+    {
+        public ulong        TimePoint;
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Services/Time/Clock/ClockTypes.cs]:
```suggestion
        public ulong   TimePoint;
        public UInt128 ClockSourceId;
```

File: Ryujinx.HLE/HOS/Services/Time/Clock/ClockTypes.cs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        public UInt128      ClockSourceId;
+    }
+
+    [StructLayout(LayoutKind.Sequential)]
+    public struct SystemClockContext
+    {
+        public ulong                Offset;
+        public SteadyClockTimePoint SteadyTimePoint;
+    }
+
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Services/Time/Clock/ClockTypes.cs]: Remove extra empty lines

File: Ryujinx.HLE/HOS/Services/Time/Clock/SteadyClockCore.cs [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+            SteadyClockTimePoint result = new SteadyClockTimePoint
+            {
+                TimePoint     = 0,
+                ClockSourceId = _clockSourceId
+            };
+
+            TimeSpanType ticksTimeSpan = TimeSpanType.FromTicks(thread.Context.ThreadState.CntpctEl0, thread.Context.ThreadState.CntfrqEl0);
+
+            result.TimePoint = _internalOffset.ToSeconds() + ticksTimeSpan.ToSeconds();
+            return result;
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Services/Time/Clock/SteadyClockCore.cs]: Add new line before

File: Ryujinx.HLE/HOS/Services/Time/IStaticService.cs [hunk 5]
```diff
@@ -1,25 +1,46 @@
...
 
+using static Ryujinx.HLE.HOS.ErrorCode;
+
 namespace Ryujinx.HLE.HOS.Services.Time
 {
-    [Service("time:a")]
-    [Service("time:s")]
-    [Service("time:u")]
+    [Flags]
+    enum TimePermissions
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Services/Time/IStaticService.cs]: Maybe you should put this enum in a `TimePermissions.cs` file

File: Ryujinx.HLE/HOS/Services/Time/IStaticService.cs [hunk 6]
```diff
@@ -1,25 +1,46 @@
...
+    {
+        LocalSystemClockWritableMask   = 0x1,
+        UserSystemClockWritableMask    = 0x2,
+        NetworkSystemClockWritableMask = 0x4,
+        UnknwonPermissionMask          = 0x8,
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Services/Time/IStaticService.cs]:
```suggestion
        UnknownPermissionMask          = 0x8,
```

File: Ryujinx.HLE/HOS/Services/Time/ISteadyClock.cs [hunk 7]
```diff
@@ -1,26 +1,18 @@
...
 namespace Ryujinx.HLE.HOS.Services.Time
 {
     class ISteadyClock : IpcService
     {
-        private ulong _testOffset;
-
-        public ISteadyClock()
-        {
-            _testOffset = 0;
-        }
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Services/Time/ISteadyClock.cs]: Remove extra empty line

File: Ryujinx.HLE/HOS/Services/Time/ISystemClock.cs [hunk 8]
```diff
@@ -1,97 +1,112 @@
...
+            _clockCore       = clockCore;
+            _writePermission = writePermission;
         }
 
         [Command(0)]
         // GetCurrentTime() -> nn::time::PosixTime
         public long GetCurrentTime(ServiceCtx context)
         {
-            DateTime currentTime = DateTime.Now;
+            SteadyClockCore steadyClockCore = _clockCore.GetSteadyClockCore();
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Services/Time/ISystemClock.cs]: Align them
```suggestion
            SteadyClockCore      steadyClockCore  = _clockCore.GetSteadyClockCore();
            SteadyClockTimePoint currentTimePoint = steadyClockCore.GetCurrentTimePoint(context.Thread);
```

File: Ryujinx.HLE/HOS/Services/Time/ISystemClock.cs [hunk 9]
```diff
@@ -1,97 +1,112 @@
...
+            if (!_writePermission)
             {
-                currentTime = currentTime.ToUniversalTime();
+                return MakeError(ErrorModule.Time, TimeError.PermissionDenied);
             }
 
-            _timeOffset = (context.RequestData.ReadInt64() - (long)(currentTime - Epoch).TotalSeconds);
+            ulong posixTime = context.RequestData.ReadUInt64();
+
+            SteadyClockCore steadyClockCore = _clockCore.GetSteadyClockCore();
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Services/Time/ISystemClock.cs]: same as `[Command(0)]`

File: Ryujinx.HLE/HOS/Services/Time/Clock/ClockTypes.cs [hunk 10]
```diff
@@ -0,0 +1,41 @@
+﻿using Ryujinx.HLE.Utilities;
+using System;
+using System.Runtime.InteropServices;
+
+namespace Ryujinx.HLE.HOS.Services.Time.Clock
+{
+    [StructLayout(LayoutKind.Sequential)]
+    public struct TimeSpanType
```
Comment (gdkchan) [inline Ryujinx.HLE/HOS/Services/Time/Clock/ClockTypes.cs]: Those structs shouldn't be public.

**Comment by Thog** [comment]
Rebased

File: Ryujinx.HLE/HOS/Services/Time/Clock/StandardUserSystemClockCore.cs [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+            }
+
+            return result;
+        }
+
+        public ResultCode SetAutomaticCorrectionEnabled(KThread thread, bool autoCorrectionEnabled)
+        {
+            ResultCode result = ApplyAutomaticCorrection(thread, autoCorrectionEnabled);
+
+            if (result == 0)
```
Comment (gdkchan) [inline Ryujinx.HLE/HOS/Services/Time/Clock/StandardUserSystemClockCore.cs]: `if (result == ResultCode.Success)`

**Pull request merged**